api/src/main/java/io/minio/encryption/MasterKey.java

```diff
@@ -0,0 +1,33 @@
+/*
+ * Minio Java SDK for Amazon S3 Compatible Cloud Storage,
+ * (C) 2017 Minio, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.minio.encryption;
+
+import java.security.InvalidKeyException;
+import java.security.NoSuchAlgorithmException;
+
+import javax.crypto.BadPaddingException;
+import javax.crypto.IllegalBlockSizeException;
+import javax.crypto.NoSuchPaddingException;
+
+public abstract class MasterKey {
+
+  protected abstract byte[] encrypt(byte[] plainText) throws IllegalBlockSizeException, BadPaddingException;
+
+  protected abstract byte[] decrypt(byte[] cipherText) throws NoSuchAlgorithmException, NoSuchPaddingException,
+      InvalidKeyException, IllegalBlockSizeException, BadPaddingException;
```

Comment:
I would prefer below in `Crypto.java`
```java
public static byte[] encrypt(byte[] plainTextKeys, KeyPair key) throws IllegalBlockSizeException, BadPaddingException;
public static byte[] encrypt(byte[] plainTextKeys, SecretKey key) throws IllegalBlockSizeException, BadPaddingException;

public static byte[] decrypt(byte[] cipherTextKeys, KeyPair key) throws NoSuchAlgorithmException, NoSuchPaddingException,
      InvalidKeyException, IllegalBlockSizeException, BadPaddingException;
public static byte[] decrypt(byte[] cipherTextKeys, SecretKey key) throws NoSuchAlgorithmException, NoSuchPaddingException,
      InvalidKeyException, IllegalBlockSizeException, BadPaddingException;

public InputStream decryptInputStream(InputStream encryptedInputStream, KeyPair key, String contentKey, byte[] iv)
      throws InvalidKeyException, NoSuchAlgorithmException, NoSuchPaddingException, IllegalBlockSizeException,
      BadPaddingException, InvalidAlgorithmParameterException;
public InputStream decryptInputStream(InputStream encryptedInputStream, SecretKey key, String contentKey, byte[] iv)
      throws InvalidKeyException, NoSuchAlgorithmException, NoSuchPaddingException, IllegalBlockSizeException,
      BadPaddingException, InvalidAlgorithmParameterException;

public InputStream encryptInputStream(InputStream plainInputStream, KeyPair key, String contentKey, byte[] iv)
      throws InvalidKeyException, NoSuchAlgorithmException, NoSuchPaddingException, IllegalBlockSizeException,
      BadPaddingException, InvalidAlgorithmParameterException;
public InputStream encryptInputStream(InputStream plainInputStream, SecretKey key, String contentKey, byte[] iv)
      throws InvalidKeyException, NoSuchAlgorithmException, NoSuchPaddingException, IllegalBlockSizeException,
      BadPaddingException, InvalidAlgorithmParameterException;
```

This would be plain/simple/straightforward methods without duplicate logic than interface/inheritance.